sensing/ahrs_ekf_mocap.hpp

```diff
@@ -0,0 +1,124 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+/*******************************************************************************
+ * \file ahrs_ekf_mocap.hpp
+ *
+ * \author MAV'RIC Team
+ * \author Matthew Douglas
+ *
+ * \brief Extended Kalman Filter attitude estimation, mixing accelerometer and magnetometer
+ * x[0] : bias_x
+ * x[1] : bias_y
+ * x[2] : bias_z
+ * x[3] : q0
+ * x[4] : q1
+ * x[5] : q2
+ * x[6] : q3
+ *
+ *
+ * Takes into account the motion capture ahrs quaternion
+ *
+ ******************************************************************************/
+
+#ifndef __AHRS_EKF_MOCAP_HPP__
+#define __AHRS_EKF_MOCAP_HPP__
+
+#include "sensing/ahrs_ekf.hpp"
+#include "util/matrix.hpp"
+#include "communication/mavlink_communication.hpp"
+
+extern "C"
+{
+}
+
+
+/**
+ * \brief The AHRS EKF class
+ */
+class Ahrs_ekf_mocap
+{
+public:
+
+    /**
+     * \brief The AHRS EKF MOCAP config structure
+     */
+    struct conf_t
+    {
+        float R_mocap;                                  ///< The variance of the accelerometer
+    };
+
+    /**
+     * \brief   AHRS EKF controller
+     *
+     * \param   message_handler The Mavlink message handler
+     * \param   ahrs_ekf        The reference to the ahrs_ekf object
+     * \param   config          Configuration structure for the mocap
+     */
+    Ahrs_ekf_mocap(Mavlink_message_handler& message_handler, Ahrs_ekf& ahrs_ekf, const conf_t config_ = Ahrs_ekf_mocap::default_config());
+
+    /**
+     * \brief   Default configuration structure
+     */
+    static inline conf_t default_config();
+
+
+protected:
+
+    /**
+     * \brief   Method used to update internal state when a message is received
+     *
+     * \param   ahrs_ekf_mocap  The ahrs_ekf_mocap object
+     * \param   sysid           ID of the system
+     * \param   msg             Pointer to the incoming message
+     */
+    static void callback(Ahrs_ekf_mocap* ahrs_ekf_mocap, uint32_t sysid, mavlink_message_t* msg);
+
+    Mat<7,1>& x_;       ///< State
+    Mat<7,7>& P_;       ///< State covariance
+    const Mat<7,7> I_;  ///< Identity matrix
+    Mat<4,4> R_mocap_;  ///< The mocap measurement noise matrix
+
```

Comment:
It seems a bit convoluted to take references to some matrices of ahrs_ekf, and to reallocate memory for an identity matrix that is already in the ahrs_ekf class.

I would have taken a reference to ahrs_ekf. Kept only R and H in this class. Then in the callback simply done:
`ahrs_ekf_.update(z, H, R);`